### PR TITLE
Symantec DLP v2

### PR DESCRIPTION
**Pull request by barchen1**
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/25659

## Description
Symantec DLP v2

## Screenshots
Screen Shot 2022-03-14 at 12 47 47
Screen Shot 2022-03-14 at 14 35 11
Screen Shot 2022-03-14 at 14 35 36
Screen Shot 2022-03-14 at 14 37 49

## Minimum version of Cortex XSOAR
- [x] 6.0.0

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+        data.update({'severity': severity})
+    custom_attributes_list = build_custom_attributes_update(custom_attributes)  # type: ignore
+    if custom_attributes_list:
+        data.update({'incidentCustomAttributes': custom_attributes_list})
+    return data
+
+
+def build_custom_attributes_update(custom_attributes: List[str]):
+    """
+    Builds the custom_attributes_list that the user wants to update.
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:547]: let's enhance the docstring here

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 2]
```diff
@@ -0,0 +1,909 @@
...
+        ),
+        outputs_prefix='SymantecDLP.IncidentStatus',
+        outputs_key_field='id',
+        outputs=incidents_status_result,
+    )
+
+
+def get_incident_history_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    incident_id = arg_to_number(args.get('incident_id'))
+    limit = arg_to_number(args.get('limit', 50))
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:725]: How big can the number of results be, in theory?

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 3]
```diff
@@ -0,0 +1,909 @@
...
+    :param last_run: Cortex XSOAR last run object
+    :param incident_types: The incident type to filter.
+    :param incident_status_id: The incident status ID to filter.
+    :param incident_severities: The incident severities to filter.
+    :return: A list of Cortex XSOAR incidents
+    """
+    if last_run and last_run.get('last_incident_creation_time'):
+        last_update_time = last_run.get('last_incident_creation_time')
+        last_incident_id = last_run.get('last_incident_id', '')
+
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:793]: No space between the if and the else please, it's one block

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 4]
```diff
@@ -0,0 +1,909 @@
...
+def get_severity_name_by_id(severity: Optional[int]):
+    """Returns the name of the severity according to the given severity ID
+
+    :param severity: The severity ID.
+
+    """
+    for severity_name, severity_num in INCIDENT_SEVERITY_MAPPING.items():
+        if severity_num == severity:
+            return severity_name
+    return ''
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:228]: Can that final return statement ever happen or is it unreachable code?

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+        # In first run
+        last_update_time = parse_creation_date(fetch_time)
+        last_incident_id = ''
+
+    incidents = []
+    if incident_severities:
+        incident_severities = [INCIDENT_SEVERITY_MAPPING[severity] for severity in incident_severities]  # type: ignore
+    if incident_types:
+        incident_types = [INCIDENT_TYPE_MAPPING[incident_type] for incident_type in incident_types]
+    incidents_data_list = client.get_incidents_request(creation_date=last_update_time, status_id=incident_status_id,
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:804]: Let's improve the variable's name

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 6]
```diff
@@ -0,0 +1,909 @@
...
+    else:
+        # In first run
+        last_update_time = parse_creation_date(fetch_time)
+        last_incident_id = ''
+
+    incidents = []
+    if incident_severities:
+        incident_severities = [INCIDENT_SEVERITY_MAPPING[severity] for severity in incident_severities]  # type: ignore
+    if incident_types:
+        incident_types = [INCIDENT_TYPE_MAPPING[incident_type] for incident_type in incident_types]
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:803]: let's add an empty line after each if block

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 7]
```diff
@@ -0,0 +1,909 @@
...
+    incidents_data_list = client.get_incidents_request(creation_date=last_update_time, status_id=incident_status_id,
+                                                       severity=incident_severities,  # type: ignore
+                                                       incident_type=incident_types, limit=fetch_limit, order_by="ASC")
+    incidents_data_list = incidents_data_list.get('incidents', [])
+    if incidents_data_list:
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:808]: Why is this line necessary? can `incidents_data_list` be `None` or just empty?

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 8]
```diff
@@ -0,0 +1,909 @@
...
+
+        for incident_data in incidents_data_list:
+            incident_id = incident_data.get('incidentId')
+            incident_creation_time = (dateparser.parse(incident_data.get('creationDate', ''))).strftime(
+                '%Y-%m-%dT%H:%M:%SZ')
+            if is_incident_already_fetched_in_previous_fetch(last_update_time, incident_creation_time, last_incident_id,
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:814]: When and why would that happen?

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 9]
```diff
@@ -0,0 +1,909 @@
...
+            (last_incident_id and int(last_incident_id) >= int(incident_id)):
+        return True
+    return False
+
+
+def fetch_incidents(client: Client, fetch_time: str, fetch_limit: int, last_run: dict, incident_types: List[str] = None,
+                    incident_status_id: List[str] = None, incident_severities: List[str] = None):
+    """
+    Performs the fetch incidents functionality, which means that every minute if fetches incidents
+    from Symantec DLP and uploads them to Cortex XSOAR server.
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:779]: Add a more detailed explanation about how the fetch works and why.
For example, why we might have already fetched incidents found more than once.
Non trivial knowledge that can't be reflected in the code should be recorded in the docstring

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 10]
```diff
@@ -0,0 +1,909 @@
...
+        if not attribute_index.isdigit():
+            raise DemistoException('Error: The attribute index must be an integer.')
+        attribute_value = splitted_att[1]
+        custom_attributes_list.append({"columnIndex": int(attribute_index), "value": attribute_value})
+    return custom_attributes_list
+
+
+def get_incident_details_fetch(client, incident):
+    """
+    Create incident details dict for each incident pulled from the fetch
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:566]: Document the 401 issue in the docstring with a detailed explanation

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 11]
```diff
@@ -0,0 +1,909 @@
...
+    """
+    incident_details = {}
+    try:
+        incident_id = incident.get('incidentId')
+        incident_type = incident.get('messageSource', '')
+        static_attributes = client.get_incident_static_attributes_request(incident_id)
+        editable_attributes = client.get_incident_editable_attributes_request(incident_id)
+        incident_details = get_common_incident_details(static_attributes, editable_attributes,
+                                                       args={"custom_attributes": "all"}, isfetch=True,
+                                                       incident_type=incident_type)
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:576]: We see these 3 lines in more than one location, let's convert them to an external function called `get_full_incident_details` or some better name, with a suitable docstring

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 12]
```diff
@@ -0,0 +1,909 @@
...
+    # In case of getting 401 (Unauthorized incident) - will get missing data
+    except DemistoException as e:
+        if '401' in str(e):
+            incident_details = get_details_unauthorized_incident(incident)
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:580]: else?

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 13]
```diff
@@ -0,0 +1,909 @@
...
+
+    :type client: ``Client``
+    :param Client: client to use
+
+    :return: 'ok' if test passed, anything else will fail the test.
+    :rtype: ``str``
+    """
+    """
+    Performs basic get request to get item samples
+    """
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:602]: 1. remove the docstring please
2. Let's make sure we also check the validity of params in the test module

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 14]
```diff
@@ -0,0 +1,909 @@
...
+            incident_details = get_incident_details_fetch(client, incident_data)
+            incident: dict = {
+                'rawJSON': json.dumps(incident_details),
+                'name': f'Symantec DLP Incident ID {incident_id}',
+                'occurred': incident_creation_time
+            }
+            incidents.append(incident)
+            if incident_id == incidents_data_list[-1].get('incidentId'):
+                last_update_time = incident_creation_time
+                last_incident_id = incident_id
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:828]: Why do we need both the time and the ID if we only compare the ID?
Let's discuss this logic as a whole flow

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+- display: Incident type
+  name: incidentType
+  options:
+  - Network
+  - Discover
+  - Endpoint
+  required: false
+  type: 16
+- additionalinfo: The status ID of the incidents. In order to get status IDs, you
+    should run the command `symantec-dlp-list-incident-status`
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml:33]: let's discuss

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml [hunk 16]
```diff
@@ -0,0 +1,417 @@
...
+  required: false
+  type: 16
+- display: Incident Severity
+  hidden: false
+  name: incidentSeverity
+  options:
+  - Info
+  - Low
+  - Medium
+  - High
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml:46]: No `additionalinfo`?

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml [hunk 17]
```diff
@@ -0,0 +1,417 @@
...
+  type: 8
+description: Symantec Data Loss Prevention version 15.7 enables you to discover, monitor and protect
+  your sensitive corporate information.
+display: Symantec Data Loss Prevention v2
+name: Symantec Data Loss Prevention v2
+script:
+  commands:
+  - arguments:
+    - default: false
+      description: Get incidents with creation date later than specified. Support
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.yml:69]:
```suggestion
      description: The earliest time from which to get incidents. Supports
```

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py [hunk 18]
```diff
@@ -0,0 +1,522 @@
...
+
+    mocker.patch.object(client, 'get_list_remediation_status_request', return_value=mock_response)
+
+    incidents_response = get_list_remediation_status(client)
+    expected_response = util_load_json('test_data/remediation_status_context.json')
+    assert incidents_response.outputs == expected_response
+
+
+@freeze_time("2022-03-07T13:34:14Z")
+def test_fetch_incidents__single(requests_mock):
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py:360]: Why the double underscore in the name, here and everywhere?

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py [hunk 19]
```diff
@@ -0,0 +1,522 @@
...
+
+    client = Client(base_url="https://SymantecDLPV2.com", auth=("test", "pass"), verify=False, proxy=False,
+                    headers={"Content-type": "application/json"})
+
+    incidents = fetch_incidents(client, fetch_time='3 days', fetch_limit='1', last_run=last_run)
+    assert len(incidents) == 1
+    assert incidents[0].get('rawJSON') == SECOND_INCIDENT_DETAILS
+
+
+def test_fetch_incidents__last_run_no_fetch(requests_mock):
```
Comment (guykeller) [inline Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py:465]: Let's discuss the test cases for the fetch mechanism

**Review by guykeller — CHANGES_REQUESTED** [review]

**Comment by barchen1** [comment]
@jrosenpanw Can you please go over the docs files?
Thanks!

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py (hunk 4 above)
Comment (guykeller) [reply Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:234]: We said we'll remove this

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py (hunk 18 above)
Comment (guykeller) [reply Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2_test.py]: change it please

**Comment by richardbluestone** [comment]
I've approved the docs. I couldn't edit this one (only view) but it looks ok.
image

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 20]
```diff
@@ -0,0 +1,909 @@
...
+    :param fetch_time: For the first time the integration is enabled with the fetch incidents functionality, the fetch
+    time indicates from what time to start fetching existing incidents in Symantec DLP system.
+    :param fetch_limit: Indicates how many incidents to fetch every minute
+    :param last_run: Cortex XSOAR last run object
+    :param incident_types: The incident type to filter.
+    :param incident_status_id: The incident status ID to filter.
+    :param incident_severities: The incident severities to filter.
+    :return: A list of Cortex XSOAR incidents
+    """
+    if last_run and last_run.get('last_incident_creation_time'):
```
[the remark this answers is not in the document]
Comment (barchen1) [reply Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:799]: no, changed

**Comment by barchen1** [comment]
@guykeller @meital-aizen I changed the fetch-incidents function because of the issue we talked about. (Incident's ID and creation date are not synchronised)
I added a note about it in the README file as well.
Let me know if there is anything you would like me to add/change.

File: Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py [hunk 21]
```diff
@@ -0,0 +1,923 @@
...
+
+
+def is_incident_already_fetched_in_previous_fetch(last_update_time, incident_creation_date):
+    """
+    Checks if the incident was already fetched
+    :param last_update_time: last_update_time from last_run
+    :param incident_creation_date: The current incident creation date
+
+    """
+    return last_update_time and last_update_time >= incident_creation_date
```
[the remark this answers is not in the document]
Comment (barchen1) [reply Packs/SymantecDLP/Integrations/SymantecDLPV2/SymantecDLPV2.py:779]: our assumption is that incidents are not created at the exact time (even if it's one millisecond difference).
@meital-aizen FYI

**Review by guykeller — APPROVED** [review]

**Comment by barchen1** [comment]
@yuvalbenshalom We will need a force merge here - error on SDK-validate
Screen Shot 2022-04-10 at 12 08 01

FYI @dorschw

**Pull request merged**